README.md

```diff
@@ -0,0 +1,111 @@
+
+# Conditions
+
+Provides:
+
+* `Condition` type as specified in the [Kubernetes API Conventions](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md)
+* `ConditionType` and generally useful constants for this type (ie. "Available",
+    "Progressing", "Degraded", and "Upgradeable")
+* Functions for setting, removing, finding, and evaluating conditions.
+
+To use, simply add `Conditions` to your Custom Resource Status struct like:
+
+```
+// ExampleAppStatus defines the observed state of ExampleApp
+type ExampleAppStatus struct {
+  ...
+  // conditions describes the state of the operator's reconciliation functionality.
+  // +patchMergeKey=type
+  // +patchStrategy=merge
+  // +optional
+  // Conditions is a list of conditions related to operator reconciliation
+  Conditions []conditions.Condition `json:"conditions,omitempty"  patchStrategy:"merge" patchMergeKey:"type"`
+}
+```
+
+Then, as appropriate in your Reconcile function, use
+`conditions.SetStatusConditions` like:
+
+```
+instance := &examplev1alpha1.ExampleApp{}
+err := r.client.Get(context.TODO(), request.NamespacedName, instance)
+...handle err
+
+conditions.SetStatusCondition(&instance.Status.Conditions, conditions.Condition{
+  Type:   conditions.ConditionAvailable,
+  Status: corev1.ConditionFalse,
+  Reason: "ReconcileStarted",
+  Message: "Reconciling resource"
+})
+
+// Update the status
+err = r.client.Status().Update(context.TODO(), instance)
+...handle err
+```
+
+
+# Object References
+
+The `ObjectReference` type is provided by Kubernetes Core API
+`"k8s.io/api/core/v1"` but the functions to set and find an `ObjectReference`
+are provided. This is useful if you would like to include in the Status of your
+Custom Resource a list of objects that are managed by your operator (ie.
+Deployments, Services, other Custom Resources, etc.).
+
+For example, we can add `RelatedObjects` to our Custom Resource Status like:
+
+```
+// ExampleAppStatus defines the observed state of ExampleApp
+type ExampleAppStatus struct {
+  ...
+  // RelatedObjects is a list of objects that are "interesting" or related to this operator.
+  RelatedObjects []corev1.ObjectReference `json:"relatedObjects,omitempty"`
+}
+```
+
+Then, through Reconcile, when an object we manage has been found we can add it to
+the `RelatedObjects` slice.
+
+```
+found := &someAPI.SomeObject{}
+err := r.client.Get(context.TODO(), types.NamespacedName{Name: object.Name, Namespace: object.Namespace}, found)
+...handle err
+
+objectreferences.SetObjectReference(&instance.Status.RelatedObjects, corev1.ObjectReference{
+  APIVersion: found.TypeMeta.APIVersion,
+  Kind:       found.TypeMeta.Kind,
+  Namespace:  found.Namespace,
+  Name:       found.Name,
+})
+
+// Update the status
+err = r.client.Status().Update(context.TODO(), instance)
+...handle err
+```
+
+# TODO
+
+* Add unit tests
+* Add support for adding reconciled version to the status. This would be useful
```

Comment:
When will it be unclear what version has been reconciled?  I may not understand the use case.

Reply:
Consider the case that you are an optional OLM installed operator and you have operator:v1 and operator:v2...your operands are tied to those image versions.

So when you are initially installed by someone using OLM you successfully rollout v1. Now, v2 comes along and OLM replaces your operator:v1 with operator:v2 and starts updating the operands. Your status may look like:

```
- available: true
- progressing: true
  reason: UpdateStarted
  message: "Rolling out v2"
- degraded: false
```

At this point, how do you know what version of the operands is currently available. If you had a status like:

```
reconciledVersion: v1
conditions:
- available: true
- progressing: true
  reason: UpdateStarted
  message: "Rolling out v2"
- degraded: false
```

It would be more clear where you stood. I think it would be nice to have but clearly not important enough to include in the first go-round.

Reply:
```
- available: true
- progressing: true
  reason: UpdateStarted
  message: "Rolling out v2"
- degraded: false
```
This status is interesting because looking only at the conditions it can mean multiple things:
- Operator is doing some work (heal)
- Operator is in the middle of a deployment (install)
- Operator is in the middle of an upgrade (upgrade)

In the meta-operator, currently we're saying "if an operator is doing work, the application isn't ready".  In other words, ```if progressing == true; then readiness = false```.  My point in mentioning this is that I'm not sure what tracking the version adds (other than for auditing).  Does the user need to know at all times, including during an upgrade, what version is Available?

Reply:
I'm assuming when you say `readiness` that you are referring to the readiness probe and not the `ConditionType ConditionAvailable`. 

I removed the todo since I'm going to add the `objectreferences` unit test shortly.

Reply:
Referring to the Readiness probe, yup.

Ok, will wait until those are done, then I think it's good to go.